---
title: ""
output: 
  github_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Intro

This pipline offers first trying to standartise procedure of soil 16s amplicone sequences Illumina reads processing in ARRIAM. Most operations performed by several libraries and covered in functions. There we will talk about most common way of analysis - but for tune or more details you can edit functions themselves, in [functions.R](/functions.R) or use a [Template](/Template.Rmd) for your own pipeline.

In this project, followed libraries have been used:

* [dada2](https://benjjneb.github.io/dada2/tutorial.html)
* [Biostrings](https://bioconductor.org/packages/release/bioc/html/Biostrings.html)
* [DECIPHER](https://bioconductor.org/packages/release/bioc/html/DECIPHER.html)
* [phyloseq](https://joey711.github.io/phyloseq/)
* [ggplot2](https://ggplot2.tidyverse.org/)
* [ggpubr](https://rpkgs.datanovia.com/ggpubr/)
* [ape](https://cran.r-project.org/web/packages/ape/index.html)
* [dplyr](https://dplyr.tidyverse.org/)
* [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq2.html)


---

## Libraries and functions import

Load requred libraries. Please, install them, if you don`t have it. Also, import functions, set your working directory and random seed.

In our test data, we will see at microbiomes of sandy soils. In this case, we compare sand with frost inclusion on an abandoned shooting range and self-grown sands (called Anclav).

```{r message=FALSE, warning=FALSE}
library(dada2)
library(Biostrings)
library(DECIPHER)
library(phyloseq)
library(ggplot2)
library(ggpubr)
library(ape)
library(dplyr)
library(DESeq2)

source('functions.R')
set.seed(5678)
setwd('/home/alexey/Analysis/16s-amplicon-processing/')
```


---

## Dada2 processing of sequences

For processing of data, we need specify a way to raw data, and file with metadata (information about samples). In our example, raw files are in `/raw` directory, and metadata in `metadata.csv` file of current work directory.

Functions in this module are:

#### Read metadata

`read_metadata(filename, sample.names, ...)`

Read metadata file, add future sample names to rownames

* `filename` - name of metadata file
* `sample.names` - name of columns with names, which you want to see in downstream analysis. Be sure, that they are unique
* `...` - you can pass any information to `read.csv()` function (for example, `sep=("\t")`)
* return a dataframe, rownames are from `sample.names` column

```{r}
mdat <- read_metadata('metadata.csv', "SampleID", sep = '\t')
mdat
```

#### dada2 pipeline 

`reads_to_seqtable_by_dada2(raw_files_path, trimLeft, truncLen, pool=TRUE, cores=TRUE)`

Process reads, plot quality data and show basic log by all steps (also save it to "processing.log" file)

* `raw_files_path` - source to raw .fastq.gz files
* `trimLeft` - if you want to cut primers from ends, use this variable as vector - c(len_forward, len_reverse)
* `truncLen` - specify the maximum length of reads, also use this variable as vector
* `pool` - pooling strategy. Options are `TRUE`, `"pseudo"` or `FALSE`. See [dada2](https://benjjneb.github.io/dada2/tutorial.html) manual.
* `cores` - number of cores for analysis. Use `TRUE` for all availible
* return ASV table and their abundance in samples

```{r, cache=TRUE}
seqtab <- reads_to_seqtable_by_dada2(raw_files_path = 'raw', trimLeft = c(19, 20), truncLen=c(220,180))
```

### Rename ASV table

According dada2 pipeline, default names of samples in seqtable are derived from names of raw files. In most cases, this names are useless, so there we can rename samples in a flow by specifying a column in metadata.

`rename_seqtab_by_metadata(seqtab, metadata, old.names)`

rename seqtab to names, specified in `read_metadata` step

* `seqtab` - ASV table from `reads_to_seqtable_by_dada2` function
* `metadata` - metadata dataframe from `read_metadata` function
* `old.names` - specify the column from `metadata` with names of a files. This names should be same with rownames of `seqtab`
* return ASV table with renamed samples

```{r}
seqtab2 <- rename_seqtab_by_metadata(seqtab, mdat, "Filename")
```

### Assign taxonomy

`assign_taxonomy(seqtab, set_train_path, train_set_species_path, cores = TRUE)`

Assign taxonomy by Bayesian naive classifier

* `seqtab` - ASV table from `reads_to_seqtable_by_dada2` function
* `set_train_path` - way to trained SILVA database fastas (see more in dada2 pipeline [here](https://benjjneb.github.io/dada2/tutorial.html))
* `train_set_species_path` - way to SILVA species fastas (see more in dada2 pipeline [here](https://benjjneb.github.io/dada2/tutorial.html))
* `cores` - number of cores for analysis. Use TRUE for all availible
* return taxonomy table

```{r, cache=TRUE}
taxa <- assign_taxonomy(seqtab = seqtab2, set_train_path = '/home/alexey/tax_n_refs/silva_nr_v132_train_set.fa.gz', 
                           train_set_species_path = '/home/alexey/tax_n_refs/silva_species_assignment_v132.fa.gz')
```

### Assemble phyloseq object

`assemble_phyloseq(seqtab, metadata, taxonomy, filter.organells = T, write_fasta = TRUE)`

Assemble phyloseq object from components (except tree)

* `seqtab` - ASV table from `rename_seqtab_by_metadata` function
* `metadata` - metadata dataframe from `read_metadata` function
* `taxonomy` - taxonomy from `assign_taxonomy` function
* `filter.organells` - filter all entries, attributes as "Mitochondria" or "Chloroplast". Can be `TRUE` or `FALSE`
* `write_fasta` - allows to write a fasta file of reference sequences in "refseqs.fasta". Can be `TRUE` or `FALSE`
* return phyloseq object


```{r}
ps <- assemble_phyloseq(seqtab = seqtab2, metadata = mdat, taxonomy = taxa, filter.organells = T, write_fasta = F)
ps
```


---

## Basic stats and save data

Feel free to explore the data and understand, how many taxa we have, reads per sample number and taxonomical structure. Also to save phyloseq object to file is always a good idea

```{r}
sample_names(ps) # Names of samples
sample_sums(ps) # Sum of reads per sample


tax_table(ps)[1:5, 1:4] # Taxonomy table
otu_table(ps)[1:4, 1:5] # ASV table

saveRDS(ps, "ps.RData")
ps <- readRDS("ps.RData")
```


---

## EDA

This part includes alpha- and beta-diversity, and bargraphs

#### Bargraphs

`bargraphps_object, rank, threshold=0.05)`

Draw a bargraph of relative abundance of different taxa in a dataset. Also result is a ggplot-object, so, it is possible add to result facet grid for group from metadata. Although unlimited number of possible sectors, only 21 unique colors are specified, so, use it on a small number of categories 

* `ps_object` - phyloseq-object
* `rank` - taxonomical level for drawing
* `threshold` - taxa with abundanse less than a threshold will be grouped in "less than" category
* return ggplot graph


```{r, fig.height=6, fig.width=8}
bargraph(ps, 'Phylum', 0.03)

bargraph(ps, 'Genus', 0.01) + facet_grid(~ Source, scale = 'free_x')
```

#### Alpha-diversity

`alpha_div_table(ps, metric, group)`

Calculate alpha-diversity indices for samples. Allows to pass columns from metadata

* `ps` - phyloseq object
* `metric` - group of metrics. Correct values are "Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher" or their group
* `group` - specify a column, or several columns from metadata to add to alpha diversity table
* return dataframe vith alpha-diversity indices


`plot_alpha(ps, metric, group)`

Plot specified alpha metric

* `ps` - phyloseq object
* `metric` - metric. Correct value is one from "Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"
* `group` - specify a column from metadata to group values
* return ggplot boxplot with points of exact values
 

```{r, fig.width=8}
alpha_div_table(ps, c("Observed", "Simpson", "Shannon"), "Source")


ggarrange(plot_alpha(ps, "Source", "Observed"),
          plot_alpha(ps, "Source", "Simpson"), plot_alpha(ps, "Source", "Shannon"),
          nrow = 1, ncol = 3)

```

#### Beta-diversity

`beta_plot(ps, method, distance, ...))`

Short functiot to draw beta diversity plot

* `ps` - phyloseq object
* `method` - method of ordination. Values are "PCoA", "NMDS"
* `...` - allows to pass arguments to `plot_ordination` function. Can be used for determination of color and shape
* return ggplot scatterplot with distances between samples


```{r, fig.width=8}
beta_plot(ps, 'PCoA', 'bray', color = "Filename", shape = "Source")
```


---

## Differential abundance

Here we try to find ASVs, which abundance significantly different in comparison within two groups. For that, we will use DeSEQ2 package. In this function, we perform comparison of two groups and return table of ASVs, significantly different from each other (p-adj < 0.05) alongside DeSEQ2 metrics.

`sig_table(ps_object, formula, threshold)`

Construct table of significant ASVs according DeSEQ2, merge it with abundance table

* `ps_object` - phyloseq object
* `formula` - formula ~var_name for grouping dataset (in our case - ~Source)
* `threshold` - baseMean and log2FoldChange, determined for filtering of deseq2 table. Use this variable as vector - c(baseMean, log2FoldChange)
* return dataframe of ASVs, their parameters in DeSEQ2 comparison and taxonomy


`draw_sig_table(sig_table, rank)`

Draw a plot by significant table

* `sig_table` - table of significant ASVs from `sig_table` function
* `rank` - taxonomical level of plot


```{r, fig.height=12, fig.width=8}
table <- sig_table(ps, ~Source, c(10, 2))
head(table)


draw_sig_table(table, 'Family')
```

`plot_heatmap(ps, group = "SampleID", log.transform = TRUE)`

Plot a heatmap by phyloseq object. Use minimal taxonomic level (Genus), and group samples by any category from metadata

* `ps` - phyloseq object
* `group` - group samples (by mean abundance). Column from the metadata
* `log.transform` - log-transformation of abundance. Can be `TRUE` or `FALSE`

DISCLAIMER: This function isn't perfect. There are lines with zero abundance in a graph (see below). In future I'm going to improve it :)

```{r, fig.height=12, fig.width=8}
# prune phyloseq according %sig_table% result
sig.ps <- prune_taxa(rownames(table), ps)

plot_heatmap(sig.ps, group = "SampleID", log.transform = TRUE)
```

---